helix-core/src/syntax.rs

```diff
@@ -231,6 +233,35 @@ impl<'de> Deserialize<'de> for FileType {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(untagged, rename_all = "kebab-case", deny_unknown_fields)]
+pub enum CommentTokens {
+    Mutliple(Vec<String>),
+    Single(String),
+}
```

Comment:
These untagged enums can feel a little clunky to work with. If we do a little more work during deserialization we can treat this type like a `Vec<String>` which is a nicer API when you want to iterate over the tokens or grab the first. We can also assert as deserialization-time that the Vec can't be empty:

```rust
#[derive(Debug, Serialize)]
pub struct CommentTokens(Vec<String>);

impl<'de> Deserialize<'de> for CommentTokens {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct Visitor;

        impl<'de> serde::de::Visitor<'de> for Visitor {
            type Value = Vec<String>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                write!(
                    formatter,
                    "a sequence of comment tokens or a single comment token"
                )
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Ok(vec![v.to_string()])
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: serde::de::SeqAccess<'de>,
            {
                let mut tokens = Vec::new();

                while let Some(token) = seq.next_element()? {
                    tokens.push(token);
                }

                if tokens.is_empty() {
                    Err(serde::de::Error::custom(
                        "`comment-tokens` must not be empty",
                    ))
                } else {
                    Ok(tokens)
                }
            }
        }

        deserializer.deserialize_any(Visitor).map(CommentTokens)
    }
}

impl ops::Deref for CommentTokens {
    type Target = Vec<String>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

Reply:
`BlockCommentTokens` can use this same pattern but instead of a `visit_str` it will want a `visit_map` that delegates to the MapAccess deserializer (which will automagically deserialize the BlockCommentToken struct)

```rust
fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
where
    A: MapAccess<'de>,
{
    let deserializer = serde::de::value::MapAccessDeserializer::new(map);
    Deserialize::deserialize(deserializer)
}
```

Reply:
Ah also looks like there's a typo, mutliple => multiple :P

Reply:
IMO its cleaner to implement the deserializer with an internal untagged enum:

``` rust 
impl<'de> Deserialize<'de> for CommentTokens {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
		#[derive(Deserialize)]
		#[serde(untagged)]
	    enum CommentTokens {
		    Mutliple(Vec<String>),
		    Single(String),
		}
		match CommentTokens::deserialize(deserializer){
			Single(val) => Self(vec![val]),
			Multile(vals) => Self(vals),
		}
	}
}
```

Since we don't really need a cust type using `#[serde(from)]` is even cleaner:

``` rust
#[derive(Deserialize)]
#[serde(untagged)]
enum CommentTokens {
    Mutliple(Vec<String>),
    Single(String),
}
impl From<CommentTokens> for Vec<String> {
    fn from(value: CommentTokens) -> Self {
        match value {
            CommentTokens::Mutliple(vals) => vals,
            CommentTokens::Single(val) => vec![val],
        }
    }
}
```

then you can just have a smile vec field

``` rust
#[serde(from = "CommentTokens"))]
comment_tokens: Vec<String>
```

Reply:
turns out ``from`` can only be used for container attributes https://serde.rs/container-attrs.html so I used ``deserialize_with`` instead

Reply:
Ah right I gorogt about that. I actually almost prefer this tough since it makes it cleat that the enun is just a deserlaize helper 👍